yrs/src/store.rs

```diff
@@ -1,58 +1,205 @@
+            }
+            _ => {
+                if let Some(inner) = self.get_type(ptr) {
+                    return Some(inner);
+                } else {
+                    None
+                }
+            }
         }
     }
-  }
-
-  pub fn get_type_ref(&self, string: &str) -> Option<u32> {
-    self.type_refs.get(string).map(|r| *r)
-  }
-  pub fn init_type_ref(&mut self, string: &str) -> u32 {
-      let types = &mut self.types;
-      *self.type_refs.entry(string.to_owned()).or_insert_with(|| {
-        let name_ref = types.len() as u32;
-        let ptr = types::TypePtr::NamedRef(name_ref);
-        let inner = types::Inner::new(ptr, None, types::TypeRefs::YArray);
-        types.push((
-          inner,
-          string.to_owned(),
-        ));
-        name_ref
-      })
-  }
-}
\ No newline at end of file
+
+    pub fn get_type_ref(&self, string: &str) -> Option<u32> {
+        self.type_refs.get(string).map(|r| *r)
+    }
+    pub fn init_type_ref(&mut self, string: &str) -> u32 {
+        let types = &mut self.types;
+        *self.type_refs.entry(string.to_owned()).or_insert_with(|| {
+            let name_ref = types.len() as u32;
+            let ptr = types::TypePtr::NamedRef(name_ref);
+            let inner = types::Inner::new(ptr, None, types::TYPE_REFS_ARRAY);
+            types.push((inner, string.to_owned()));
+            name_ref
+        })
+    }
+    pub fn create_item(&mut self, pos: &block::ItemPosition, content: block::ItemContent) {
+        let parent = self.get_type(&pos.parent).unwrap();
+        let left = pos.after;
+        let right = match pos.after.as_ref() {
+            Some(left_id) => self.blocks.get_item(left_id).right,
+            None => parent.start.get(),
+        };
+        let id = block::ID {
+            client: self.client_id,
+            clock: self.get_local_state(),
+        };
+        let pivot = self
+            .blocks
+            .get_client_blocks_mut(self.client_id)
+            .integrated_len as u32;
+        let item = block::Item {
+            id,
+            content,
+            left,
+            right,
+            origin: pos.after.as_ref().map(|l| l.id),
+            right_origin: right.map(|r| r.id),
+            parent: pos.parent.clone(),
+            deleted: false,
+            parent_sub: None,
+        };
+        item.integrate(self, pivot as u32);
+        let local_block_list = self.blocks.get_client_blocks_mut(self.client_id);
+        local_block_list.list.push(block::Block::Item(item));
+        local_block_list.integrated_len += 1;
+    }
+
+    pub fn read_blocks(&mut self, update_decoder: &mut updates::decoder::DecoderV1) {
+        let number_of_clients: u32 = update_decoder.rest_decoder.read_var_uint();
+        for _ in 0..number_of_clients {
+            let client = update_decoder.read_client();
+            let number_of_structs: u32 = update_decoder.rest_decoder.read_var_uint();
```

Comment:
Adding latest test (`apply_update_basic`) confirmed that this step doesn't work as expected. Number of structs assigned by yjs should be 3, but upon reading its 1490905955.